scripts/smt-state-migration.ts

```diff
@@ -1,23 +1,142 @@
+    stateContract: any,
+    startBlockNumber = 0, //29831814
+    blockChunkSize = 3500,
+    enableLogging = false
+  ): Promise<any[]> {
     const filter = stateContract.filters.StateUpdated(null, null, null, null);
-    const stateHistory = await stateContract.queryFilter(filter);
-    return stateHistory;
+    const latestBlock = await ethers.provider.getBlock("latest");
+    enableLogging && console.log("latestBlock Number", latestBlock.number);
+
+    let stateTransitionHistory: unknown[] = [];
+
+    for (
+      let index = startBlockNumber;
+      index <= latestBlock.number;
+      index += blockChunkSize
+    ) {
+      let pagedHistory;
+      try {
+        pagedHistory = await stateContract.queryFilter(
+          filter,
+          index,
+          index + blockChunkSize - 1
+        );
+      } catch (error) {
+        console.error(error);
+      }
+      enableLogging &&
+        console.log(
+          `state transition history length: ${pagedHistory.length}, current block number: ${index}, latest block number: ${latestBlock.number}`
+        );
+      stateTransitionHistory = [...stateTransitionHistory, ...pagedHistory];
+    }
+    enableLogging &&
+      console.log(`Total events count: ${stateTransitionHistory.length}`);
+
+    // save data to file
+    SmtStateMigration.writeFile("events-data.json", stateTransitionHistory);
+
+    return stateTransitionHistory;
   }
 
-  static async migrate(smt: any, stateHistory: any[]): Promise<void> {
-    for (let index = 0; index < stateHistory.length; index++) {
-      const [id, block, timestamp, state] = stateHistory[index].args;
-      await smt.addHistorical(id, state, timestamp, block);
+  static async migrate(
+    smtContract: any,
+    stateTransitionHistory: any[],
+    enableLogging = false
+  ): Promise<void> {
+    const result: { migratedData: any[]; error: unknown; index: number } = {
+      migratedData: [],
+      error: null,
+      index: 0,
+    };
+    for (let index = 0; index < stateTransitionHistory.length; index++) {
+      const [id, block, timestamp, state] = stateTransitionHistory[index].args;
+      result.index = index;
+      try {
+        await smtContract.addHistorical(id, state, timestamp, block);
+
+        result.migratedData.push({ id, state, timestamp, block });
+      } catch (error) {
+        result.error = error;
+        break;
+      }
+    }
+    if (!result.error) {
+      enableLogging && console.log("migration completed successfully", result);
+    } else {
+      enableLogging && console.log("migration error", result.error);
     }
+    SmtStateMigration.writeFile("migration-result.json", result);
   }
 
-  static async upgradeState(existingStateAddress: any) {
-    const stateFactory = await ethers.getContractFactory("State");
-    await upgrades.upgradeProxy(existingStateAddress, stateFactory);
-    const stateContract = stateFactory.attach(existingStateAddress);
+  static async upgradeState(
+    existingStateAddress: string,
```

Comment:
Something like `stateProxyAddress` may be more precise name than `existingStateAddress`.

The same goes for any place in the code, where this method is called: if variable that is assigned to this param is `existingStateAddress` then it is better to change it.